docs/concepts/filters.md

```diff
@@ -0,0 +1,307 @@
+---
+title: Filters
+author: halter73
+description: MCP Server Handler Filters
+uid: filters
+---
+
+# MCP Server Handler Filters
+
+For each handler type in the MCP Server, there are corresponding `AddXXXFilter` methods in `McpServerBuilderExtensions.cs` that allow you to add filters to the handler pipeline. The filters are stored in `McpServerOptions.Filters` and applied during server configuration.
+
+## Available Filter Methods
+
+The following filter methods are available:
+
+- `AddListResourceTemplatesFilter` - Filter for list resource templates handlers
+- `AddListToolsFilter` - Filter for list tools handlers
+- `AddCallToolFilter` - Filter for call tool handlers
+- `AddListPromptsFilter` - Filter for list prompts handlers
+- `AddGetPromptFilter` - Filter for get prompt handlers
+- `AddListResourcesFilter` - Filter for list resources handlers
+- `AddReadResourceFilter` - Filter for read resource handlers
+- `AddCompleteFilter` - Filter for completion handlers
+- `AddSubscribeToResourcesFilter` - Filter for resource subscription handlers
+- `AddUnsubscribeFromResourcesFilter` - Filter for resource unsubscription handlers
+- `AddSetLoggingLevelFilter` - Filter for logging level handlers
```

Comment:
Did you consider any kind of global filter that would enable intercepting arbitrary requests/responses, regardless of method?

Reply:
I considered it briefly, but I wasn't sure there was enough in common for each type of request to be that useful. If we want to do this, we'd probably want to add a base type for `RequestContext<TParams>` that the global filter could accept. I think this could be done as a follow up.

Reply:
I think a global transport-level filter that could see all incoming and outgoing JsonRpcMessages would be useful for both the server and client. I don't think these filters would replace the filters added here since it would be much more difficult to implement something like an authorization filter that early in the pipeline, but I think it could be particularly useful for diagnostics on both the server and client. I filed an issue tracking that kind of global transport-level filter at #761.